feat(dafny): Add bucket beacon support

ajewellamz opened the pull request:
*Issue #, if available:*

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

ajewellamz committed:
chore(dafny): Add bucket beacon support

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
new smithy-dafny

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

mahnushm committed:
Update background.md

Small editorial errors.

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'ajewell/buckets' of github.com:aws/aws-database-encryption-sdk-dynamodb into ajewell/buckets

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
rename bucket to partition

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

ajewellamz committed:
m

ajewellamz committed:
Merge branch 'ajewell/buckets' of github.com:aws/aws-database-encryption-sdk-dynamodb into ajewell/buckets

ajewellamz committed:
Merge branch 'main' into ajewell/buckets

lucasmcdonald3 committed:
Merge branch 'main' into ajewell/buckets

rishav-karanjit committed:
Merge branch 'main' into ajewell/buckets

rishav-karanjit commented on specification/changes/2025-08-25-beacon-partitions/background.md:
```diff
@@ -0,0 +1,321 @@
+while still guaranteeing that, given the partition for the item,
+we can uniquely identify the partition for the constrained beacon.
+
+**_WARNING_** Adding or changing the constraint on a beacon is difficult, sometimes impossible;
```
Praise: Thank you are highlighting it. This is a easy miss for a first time reader.

rishav-karanjit commented on TestVectors/dafny/DDBEncryption/src/TestVectors.dfy:
```diff
@@ -128,6 +168,704 @@ module {:options "-functionSyntax:4"} DdbEncryptionTestVectors {
+        AttrNames[6] := AttrValues[6],
+        "PreferredPartition" := DDB.AttributeValue.N(num2)
+      ]
+    }
```
nit:

```suggestion
    function MakePartitionRecord(x : nat) : DDB.AttributeMap
    {
      var recordNumber := String.Base10Int2String(x);
      var partitionNumber := String.Base10Int2String(x%5);

      map[
        HashName := DDB.AttributeValue.N(recordNumber),
        AttrNames[0] := AttrValues[0],
        AttrNames[1] := AttrValues[1],
        AttrNames[2] := AttrValues[2],
        AttrNames[3] := AttrValues[3],
        AttrNames[4] := AttrValues[4],
        AttrNames[5] := AttrValues[5],
        AttrNames[6] := AttrValues[6],
        "PreferredPartition" := DDB.AttributeValue.N(partitionNumber)
      ]
    }
```

rishav-karanjit commented on TestVectors/dafny/DDBEncryption/src/TestVectors.dfy:
```diff
@@ -128,6 +168,704 @@ module {:options "-functionSyntax:4"} DdbEncryptionTestVectors {
+          print "Partition Scan ", queryName, " did not find record ", i, "\n";
+          wasBad := true;
+        } else if counts[i] != 1 {
+          print "Partition Scab ", queryName, " returned record ", i, " ", counts[i], " times\n";
```
nit:
```suggestion
          print "Partition Scan ", queryName, " returned record ", i, " ", counts[i], " times\n";
```

rishav-karanjit reviewed (commented):
I have reviewed everything under specification and testVectors. There is one blocking comment in spec. Remaining are suggestions but not blocking. 

I will review rest the next time.

mahnushm committed:
Update specification/changes/2025-08-25-beacon-partitions/background.md

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

mahnushm committed:
Update TestVectors/dafny/DDBEncryption/src/TestVectors.dfy

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

mahnushm committed:
Update TestVectors/dafny/DDBEncryption/src/JsonConfig.dfy

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

mahnushm committed:
Update TestVectors/dafny/DDBEncryption/src/TestVectors.dfy

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

texastony committed:
Merge branch 'main' into ajewell/buckets

texastony committed:
fix: typo in test vectors

rishav-karanjit commented on DynamoDbEncryption/dafny/DynamoDbEncryption/src/DDBSupport.dfy:
```diff
@@ -221,6 +235,74 @@ module DynamoDBSupport {
+    var filterHasEncField := Filter.UsesEncryptedField(Filter.ParseExprOpt(filterExpr), actions, names);
+    var keyHasEncField := Filter.UsesEncryptedField(Filter.ParseExprOpt(keyExpr), actions, names);
+    if keyHasEncField.Some? || filterHasEncField.Some? {
+      return Failure(E("When numberOfPartitions is greater than one, XXXValues must contain " + PartitionName));
```
"XXXValues" in the error message seems to be a typo?

rishav-karanjit replied:
```suggestion
      return Failure(E("When numberOfPartitions is greater than one, Values must contain " + PartitionName));
```

rishav-karanjit reviewed (commented):
I am few comments/question

mahnushm committed:
Updated example in java to use partitions

mahnushm committed:
Fix partition variable usage in query request

mahnushm committed:
Fix break statement in BasicSearchableEncryptionExample

mahnushm committed:
Refactor comments and update beacon configurations

mahnushm committed:
Refactor query response handling in basic_searchable_encryption

mahnushm committed:
Update basic_searchable_encryption.rs

mahnushm committed:
Add partition condition to query request

mahnushm committed:
Update BasicSearchableEncryptionExample.java

Minor error

mahnushm committed:
Fix method chaining in beacon builder and update query

mahnushm committed:
Update basic_searchable_encryption.rs

mahnushm committed:
Update BasicSearchableEncryptionExample.java

mahnushm committed:
Remove partition key from query condition expression

mahnushm committed:
Update basic_searchable_encryption.rs

mahnushm committed:
Update BasicSearchableEncryptionExample.java

Adding GetNumberOfQueries and defaultNumberOfPartitions.

mahnushm committed:
Update BasicSearchableEncryptionExample.java

Change i to partition

mahnushm committed:
Update basic_searchable_encryption.rs

Adding default_number_of_partitions(1)

mahnushm committed:
Update basic_searchable_encryption.rs

Adding get_number_of_queries logic.

mahnushm committed:
Update BasicSearchableEncryptionExample.cs

Adding partitions to the breacons.

mahnushm committed:
Update BasicSearchableEncryptionExample.java

Adding updatedQueryRequest

mahnushm committed:
Update BasicSearchableEncryptionExample.java

mahnushm committed:
Update Examples/runtimes/java/DynamoDbEncryption/src/main/java/software/amazon/cryptography/examples/searchableencryption/BasicSearchableEncryptionExample.java

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

mahnushm committed:
Update Examples/runtimes/java/DynamoDbEncryption/src/main/java/software/amazon/cryptography/examples/searchableencryption/BasicSearchableEncryptionExample.java

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

mahnushm committed:
Update basic_searchable_encryption.rs

mahnushm committed:
Update BasicSearchableEncryptionExample.java

mahnushm committed:
Update BasicSearchableEncryptionExample.java

Adding correct usage of DynamoDbTablesEncryptionConfig

mahnushm committed:
Refactor DynamoDB encryption configuration

Removed unnecessary table configuration setup for DynamoDB.

mahnushm committed:
Clarify numQueries usage with additional comments

Add comments to clarify the purpose of numQueries.

mahnushm committed:
Fix DynamoDbTablesEncryptionConfig initialization

mahnushm committed:
Change numQueries from 1 to 2

mahnushm committed:
Update BasicSearchableEncryptionExample.java

Testing a value that is not power of 2

mahnushm committed:
Modify partition settings in BasicSearchableEncryptionExample

Updated maximumNumberOfPartitions to 4 and clarified defaultNumberOfPartitions constraints.

mahnushm committed:
Refactor numQueries retrieval using queryInputTransform

mahnushm committed:
Modify beacon partitioning and query handling

Updated number of partitions for beacons and adjusted related logic to ensure correct handling of query results.

mahnushm committed:
Rename logicalTableName to tableName in transform input

mahnushm committed:
Refactor query input transformation and number retrieval

mahnushm committed:
Update BasicSearchableEncryptionExample.java

mahnushm commented on specification/changes/2025-08-25-beacon-partitions/change.md:
```diff
@@ -0,0 +1,92 @@
+
+- An integer in the range 0..254
+- The number of the partition currently under consideration in some context
+- For customers not using PartitionBeacons, this is always `1`
```
You are correct, it should be 0.

mahnushm committed:
Change default PartitionNumber value to 0

Updated the default value for PartitionNumber from 1 to 0 for customers not using PartitionBeacons.

mahnushm committed:
Remove unused imports in BasicSearchableEncryptionExample

Removed unused imports related to QueryInputTransform.

mahnushm committed:
Update BasicSearchableEncryptionExample.java

mahnushm committed:
Update BasicSearchableEncryptionExample.java

mahnushm committed:
Update number of partitions in StandardBeacon configuration

mahnushm committed:
Comment out numberOfPartitions in StandardBeacons

Comment out the numberOfPartitions setting for two StandardBeacons.

mahnushm committed:
Update BasicSearchableEncryptionExample.java

mahnushm committed:
Adjust partition settings in basic_searchable_encryption

josecorella committed:
test(dafny): beacon partition tests (#2199)

mahnushm committed:
Format Java file with prettier

mahnushm committed:
correct formatting

mahnushm committed:
correct formating

mahnushm committed:
Merge branch 'mahnushm-example-rust' into ajewell/buckets

mahnushm committed:
Adding java example for partitioned beacon

mahnushm committed:
Adding asynch query to help with delay.

josecorella committed:
Merge branch 'main' into ajewell/buckets

josecorella committed:
Merge branch 'main' into ajewell/buckets

josecorella committed:
format

josecorella committed:
Merge branch 'main' into ajewell/buckets

josecorella committed:
Merge branch 'main' into ajewell/buckets

lucasmcdonald3 reviewed (dismissed):
Approving on Jose's behalf to unblock merge, I haven't reviewed these changes

josecorella commented on DynamoDbEncryption/dafny/DynamoDbEncryption/Model/DynamoDbEncryption.smithy:
```diff
@@ -49,6 +49,46 @@ service DynamoDbEncryption {
+@reference(resource: PartitionSelector)
+structure PartitionSelectorReference {}
+
+operation GetPartitionNumber {
+  input: GetPartitionNumberInput,
+  output: GetPartitionNumberOutput,
+}
+
+//= specification/searchable-encryption/search-config.md#partition-selector
+//= type=implication
+//# GetPartitionNumber MUST take as input
+//#
+//# - A DynamoDB Item (i.e an AttributeMap)
+//# - The [number of partitions](#max-partitions) defined in the associated [beacon version](#beacon-version-initialization).
+//# - The logical table name for this defined in the associated [table config](../dynamodb-encryption-client/ddb-table-encryption-config.md#structure).
+
+structure GetPartitionNumberInput {
+  @required
+  item: AttributeMap,
+  @required
+  numberOfPartitions : PartitionCount,
+  @required
+  logicalTableName: String,
+}
+
+//= specification/searchable-encryption/search-config.md#partition-selector
+//= type=implication
+//# GetPartitionNumber MUST return
+//#
+//# - The number of the partition to use for this item
+
+structure GetPartitionNumberOutput {
+  @required
+  partitionNumber: PartitionNumber
+}
+
```
this operation and new structures have no javadocs annotations,should we add some @mahnushm?
(I also realize that the javadoc annotation is only supported in java afaik but it would be useful to add)

mahnushm replied:
Is it useful to have them even if it is not java?

josecorella commented on DynamoDbEncryption/dafny/DynamoDbEncryption/src/DDBSupport.dfy:
```diff
@@ -237,8 +319,14 @@ module DynamoDBSupport {
       var keyId :- Filter.GetBeaconKeyId(search.value.curr(), req.KeyConditionExpression, req.FilterExpression, req.ExpressionAttributeValues, req.ExpressionAttributeNames);
-      var oldContext := Filter.ExprContext(req.KeyConditionExpression, req.FilterExpression, req.ExpressionAttributeValues, req.ExpressionAttributeNames);
-      var newContext :- Filter.Beaconize(search.value.curr(), oldContext, keyId);
+      var foo :- ExtractPartition(search.value.curr(), req.FilterExpression, req.KeyConditionExpression, req.ExpressionAttributeNames, req.ExpressionAttributeValues, actions);
```
@rishav-karanjit, did you resolve this? does it need to be addressed before merging?

rishav-karanjit replied:
It still looks relevant to me when I looked into it today.

josecorella commented on DynamoDbEncryption/dafny/DynamoDbEncryption/src/FilterExpr.dfy:
```diff
@@ -1662,6 +1702,34 @@ module DynamoDBFilterExpr {
+    for i : uint64 := 0 to |values| as uint64
+      invariant result <= bv.numPartitions
+    {
+      var partitions := values[i].0.getNumQueries(bv.numPartitions, values[0].1);
```
did this get answered?

rishav-karanjit replied:
Still looks relevant to me when reviewed today.

josecorella commented on DynamoDbEncryption/dafny/DynamoDbEncryptionTransforms/Model/DynamoDbEncryptionTransforms.smithy:
```diff
@@ -73,6 +76,29 @@ map StringMap {
+//= specification/dynamodb-encryption-client/ddb-get-number-of-queries.md#input
+//= type=implication
+//# This operation MUST take as input the QueryInput structure under consideration.
+structure GetNumberOfQueriesInput {
+    @required
+    input: QueryInput
+}
+
+//= specification/dynamodb-encryption-client/ddb-get-number-of-queries.md#input
+//= type=implication
+//# This operation MUST return the number of queries necessary.
+structure GetNumberOfQueriesOutput {
+    @required
+    numberOfQueries: PartitionCount
+}
```
these dont have javadocs

josecorella reviewed (commented):
@mahnushm I also see that there are no rust examples, are we merging those after this PR lands but before launch?

rishav-karanjit committed:
Merge branch 'main' into ajewell/buckets

rishav-karanjit reviewed (commented):
There are two of my comments that still looks relevant to me

https://github.com/aws/aws-database-encryption-sdk-dynamodb/pull/1943/changes#r2962969885

https://github.com/aws/aws-database-encryption-sdk-dynamodb/pull/1943/changes#r2962931735

rishav-karanjit committed:
Merge branch 'main' into ajewell/buckets

rishav-karanjit commented on DynamoDbEncryption/dafny/DynamoDbEncryption/src/DDBSupport.dfy:
```diff
@@ -237,8 +319,14 @@ module DynamoDBSupport {
       var keyId :- Filter.GetBeaconKeyId(search.value.curr(), req.KeyConditionExpression, req.FilterExpression, req.ExpressionAttributeValues, req.ExpressionAttributeNames);
-      var oldContext := Filter.ExprContext(req.KeyConditionExpression, req.FilterExpression, req.ExpressionAttributeValues, req.ExpressionAttributeNames);
-      var newContext :- Filter.Beaconize(search.value.curr(), oldContext, keyId);
+      var foo :- ExtractPartition(search.value.curr(), req.FilterExpression, req.KeyConditionExpression, req.ExpressionAttributeNames, req.ExpressionAttributeValues, actions);
```
```suggestion
      var foo :- ExtractPartition(search.value.curr(), req.KeyConditionExpression, req.FilterExpression, req.ExpressionAttributeNames, req.ExpressionAttributeValues, actions);
```

rishav-karanjit commented on DynamoDbEncryption/dafny/DynamoDbEncryption/src/FilterExpr.dfy:
```diff
@@ -1662,6 +1701,34 @@ module DynamoDBFilterExpr {
+    for i : uint64 := 0 to |values| as uint64
+      invariant result <= bv.numPartitions
+    {
+      var partitions := values[i].0.getNumQueries(bv.numPartitions, values[0].1);
```
```suggestion
      var partitions := values[i].0.getNumQueries(bv.numPartitions, values[i].1);
```

mahnushm committed:
Update DynamoDbEncryption/dafny/DynamoDbEncryption/src/DDBSupport.dfy

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

mahnushm committed:
Update DynamoDbEncryption/dafny/DynamoDbEncryption/src/FilterExpr.dfy

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

mahnushm committed:
Update DynamoDbEncryption/dafny/DynamoDbEncryption/src/DDBSupport.dfy

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

rishav-karanjit reviewed (dismissed):
LGTM

rishav-karanjit commented on DynamoDbEncryption/dafny/DynamoDbEncryption/test/BeaconPartition.dfy:
```diff
@@ -0,0 +1,718 @@
+    // --- Missing partition when partitions configured and encrypted fields used (should fail) ---
+    var res3 := DDBS.ExtractPartition(bv, None, Some("Name = :val"), None, None, FullTableConfig.attributeActionsOnEncrypt);
+    expect res3.Failure?;
+    expect_equal(res3.error, E("When numberOfPartitions is greater than one, XXXValues must contain :aws_dbe_partition"));
```
```suggestion
    expect_equal(res3.error, E("When numberOfPartitions is greater than one, Values must contain :aws_dbe_partition"));
```

mahnushm committed:
Apply suggestion from @rishav-karanjit

Co-authored-by: Rishav karanjit <karanjitrishav4@gmail.com>

rishav-karanjit committed:
Merge branch 'main' into ajewell/buckets

rishav-karanjit committed:
Merge branch 'main' into ajewell/buckets
